### PR TITLE
w3c(phase 3): ensure last datadog parent id is always recorded

### PR DESCRIPTION
**Pull request by mabdinur**
**What does this PR do?**

Ensures when trace headers with the same trace_id but different span_ids are received by an application the tracecontext headers are treated as the source of truth.
- The traceparent header will be used to set the parent_id
- If the conflicting trace information was extracted from datadog headers, the span_id from the datadog headers will be stored in the `_dd.parent_id` parent_id tag. 
- All other propagation styles are third class citizens. If they conflict with tracecontext headers their span_ids will be overwritten. 

This behavior will be validated by the following system-tests:  https://github.com/DataDog/system-tests/pull/2385

**Motivation:**

Provide first class support for services instrumented with opentelemetry and reduce instances of "broken traces".

**Additional Notes:**

**How to test the change?**

Unsure? Have a question? Request a review!

### CODE REVIEW

File: lib/datadog/tracing/distributed/propagation.rb [hunk 1]
```diff
@@ -100,7 +100,14 @@ def extract(data)
             if (tracecontext_digest = propagator.extract(data))
               # Only parse if it represent the same trace as the successfully extracted one
               next unless tracecontext_digest.trace_id == extracted_trace_digest.trace_id
-
+
```
Comment (mabdinur) [inline lib/datadog/tracing/distributed/propagation.rb:103]:
```suggestion

```

File: lib/datadog/tracing/distributed/propagation.rb [hunk 2]
```diff
@@ -100,7 +100,13 @@ def extract(data)
             if (tracecontext_digest = propagator.extract(data))
               # Only parse if it represent the same trace as the successfully extracted one
               next unless tracecontext_digest.trace_id == extracted_trace_digest.trace_id
-
+              tracecontext_digest.trace_distributed_tags |= {}
```
Comment (mabdinur) [inline lib/datadog/tracing/distributed/propagation.rb:103]:
```suggestion

              tracecontext_digest.trace_distributed_tags |= {}
```

File: lib/datadog/tracing/distributed/propagation.rb [hunk 3]
```diff
@@ -101,10 +101,22 @@ def extract(data)
               # Only parse if it represent the same trace as the successfully extracted one
               next unless tracecontext_digest.trace_id == extracted_trace_digest.trace_id
 
-              # Preserve the `tracestate`
+              parent_id = extracted_trace_digest.span_id
+              distributed_tags = extracted_trace_digest.trace_distributed_tags
+              unless extracted_trace_digest.span_id == tracecontext_digest.span_id
+                # span_id in the tracecontext header takes precedence over the value in datadog headers
```
Comment (mabdinur) [inline lib/datadog/tracing/distributed/propagation.rb:107]:
```suggestion
                # span_id in the tracecontext header takes precedence over the value in all conflicting headers
```

File: lib/datadog/tracing/distributed/propagation.rb [hunk 4]
```diff
@@ -115,6 +127,22 @@ def extract(data)
 
           extracted_trace_digest
         end
+
+        private
+
+        def last_datadog_parent_id(headers, tracecontext_tags)
+          dd_propagator = @propagation_style_extract.find { |propagator| propagator.is_a?(Datadog) }
+          if tracecontext_tags&.fetch(
```
Comment (mabdinur) [inline lib/datadog/tracing/distributed/propagation.rb:135]: Add test case for this branch

**Review by marcotc — APPROVED** [review]

**Pull request merged**